File: Leetcode-C++/Leetcode-1 Two Sum.cpp
#include <vector>
#include <unordered_map>
using namespace std;

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int,int> dict;
        for (int i=0; i < nums.size(); i++) {
            int temp = target - nums[i];
            if (dict.count(temp)) {
                // c++判断是不是某个字典的key的办法: dict.count(key)
                return {i,dict[temp]};
                // c++的构成数组的方式 不可以用[]
            } else {
                dict[nums[i]]=i;
            }
        }
        return {};
    }
};

File: Leetcode-Python/Leetcode-1 Two Sum.py
class Solution(object):
    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        hashmap={}
        for i,num in enumerate(nums):
            complement=target-num
            if complement in hashmap:
                return [hashmap[complement] ,i]
            else:
                hashmap[num]=i
#注意点：哈希表的使用，在里面？输出，不在里面？存入
#注意这个哈希表它的元素是i，然后它的index才是num，因为我们要求输出的是原数列的i（index）
#enumerate的用法，同时告诉我元素和index，适用于任何iterable的东西
#哈希表的优势体现在查找的时候，复杂程度为O(1)，不像普通的查找需要一个一个去看